airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py

```diff
@@ -235,84 +259,45 @@ def execute(self, context):
-                self.pod_runtime_info_envs +
-                self.ports +
-                self.resources +
-                self.secrets +
-                self.volumes +
-                self.volume_mounts
-            )
+            label_selector = self._get_pod_identifying_label_string(labels)
 
-            self.pod = pod
+            pod_list = client.list_namespaced_pod(self.namespace, label_selector=label_selector)
 
-            launcher = pod_launcher.PodLauncher(kube_client=client,
-                                                extract_xcom=self.do_xcom_push)
+            if len(pod_list.items) > 1:
+                raise AirflowException(
+                    'More than one pod running with labels: '
+                    '{label_selector}'.format(label_selector=label_selector))
 
-            try:
-                (final_state, result) = launcher.run_pod(
-                    pod,
-                    startup_timeout=self.startup_timeout_seconds,
-                    get_logs=self.get_logs)
-            except AirflowException:
-                if self.log_events_on_failure:
-                    for event in launcher.read_pod_events(pod).items:
-                        self.log.error("Pod Event: %s - %s", event.reason, event.message)
-                raise
-            finally:
-                if self.is_delete_operator_pod:
-                    launcher.delete_pod(pod)
+            launcher = pod_launcher.PodLauncher(kube_client=client, extract_xcom=self.do_xcom_push)
 
+            if len(pod_list.items) == 1 and self._try_numbers_do_not_match(context, pod_list.items[0]):
+                launcher.delete_pod(pod_list.items[0])
+                final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
+            elif len(pod_list.items) == 1:
+                self.log.info("found a running pod with labels %s."
+                              "Will monitor instead of starting new one", labels)
+                final_state, result = self.monitor_launched_pod(launcher, pod_list[0])
+            else:
+                final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
             if final_state != State.SUCCESS:
-                if self.log_events_on_failure:
-                    for event in launcher.read_pod_events(pod).items:
-                        self.log.error("Pod Event: %s - %s", event.reason, event.message)
                 raise AirflowException(
-                    'Pod returned a failure: {state}'.format(state=final_state)
-                )
-
+                    'Pod returned a failure: {state}'.format(state=final_state))
             return result
         except AirflowException as ex:
             raise AirflowException('Pod Launching failed: {error}'.format(error=ex))
 
+    @staticmethod
+    def _get_pod_identifying_label_string(labels):
+        filtered_labels = {label_id: label for label_id, label in labels.items() if label_id != 'try_number'}
```

Comment:
Is it not possible to use multiple labels to get the pod instead of creating a single string?

If not, we need to be careful as the order of items can change how the string. We can `sort` the keys in `filtered_labels` to take care of that issue.

Reply:
@kaxil why do we need to sort? this is for creating the label_selector for kubernetes which I don't believe is order specific.

Reply:
For the following reason:

```
In [1]: def _get_pod_identifying_label_string(labels):
   ...:         filtered_labels = {label_id: label for label_id, label in labels.items() if label_id != 'try_number'}
   ...:         return ','.join([label_id + '=' + label for label_id, label in filtered_labels.items()])
   ...:

In [2]: labels = {
   ...:             'dag_id': 'test_dag_id',
   ...:             'task_id': 'test_task_id',
   ...:             'execution_date': 'execution_date',
   ...:             'try_number': 2,
   ...:         }

In [3]: _get_pod_identifying_label_string(labels)
Out[3]: 'dag_id=test_dag_id,task_id=test_task_id,execution_date=execution_date'

In [6]: labels_1 = {
   ...:             'task_id': 'test_task_id',
   ...:             'dag_id': 'test_dag_id',
   ...:             'execution_date': 'execution_date',
   ...:             'try_number': 2,
   ...:         }

In [7]: _get_pod_identifying_label_string(labels_1)
Out[7]: 'task_id=test_task_id,dag_id=test_dag_id,execution_date=execution_date'

```